core/classes/Placeholders.php

```diff
@@ -0,0 +1,162 @@
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Create a new row in nl2_placeholders_settings if a row with the "name" of $name does not exist.
+     * 
+     * @param string $name Name of placeholder
+     */
+    public function registerPlaceholder($name) {
+        $this->_db->query("INSERT IGNORE INTO nl2_placeholders_settings (name) VALUES (?)", [$name]);
+    }
+
+    /**
+     * Load placeholders for this specific user.
+     * 
+     * @param string $uuid Their valid Minecraft uuid to use for lookup.
+     * @return array Their placeholders.
+     */
+    public function loadUserPlaceholders($uuid) {
+
+        $user_placeholders = [];
+
+        // TODO: probably can handle assigning friendly_name from within the query
+        $placeholders = $this->_db->query('SELECT * FROM nl2_users_placeholders up JOIN nl2_placeholders_settings ps ON up.name = ps.name WHERE up.uuid = ?', [$uuid]);
+
+        if (!$placeholders->count()) {
+            return $user_placeholders;
+        }
+
+        $placeholders = $placeholders->results();
+        foreach ($placeholders as $placeholder) {
+            $data = new stdClass();
+
+            $data->server_id = $placeholder->server_id;
+            $data->name = $placeholder->name;
+            $data->friendly_name = isset($placeholder->friendly_name) ? $placeholder->friendly_name : $placeholder->name;
+            $data->value = $placeholder->value;
+            $data->last_updated = $placeholder->last_updated;
+            $data->show_on_profile = $placeholder->show_on_profile;
+            $data->show_on_forum = $placeholder->show_on_forum;
+
+            $user_placeholders[$placeholder->name] = $data;
+        }
+
+        return $user_placeholders;
+    }
+
+    /**
+     * Get all placeholders which are set to have leaderboards.
+     * 
+     * @return array Array of placeholders which have leaderboard enabled.
+     */
+    public function getLeaderboardPlaceholders() {
+        return array_filter($this->_all_placeholders, fn($placeholder) => $placeholder->leaderboard);
+    }
+
+    /**
+     * Get (cached) leaderboard data for a specific leaderboard.
+     * 
+     * @param string $placeholder_name Unique name of placeholder to get data for.
+     * @return array|null Array of data or null if this placeholder is not setup for leaderboards.
+     */
+    public function getLeaderboardData($placeholder_name) {
+        if (!in_array($placeholder_name, array_column($this->getLeaderboardPlaceholders(), 'name'))) {
+            return null;
+        }
+
+        if (isset($this->_leaderboard_data[$placeholder_name])) {
+            return $this->_leaderboard_data[$placeholder_name];
+        }
+
+        $sort = $this->getPlaceholderByName($placeholder_name)->leaderboard_sort;
+
+        $leaderboard_data = $this->_db->query("SELECT * FROM nl2_users_placeholders WHERE name = ? ORDER BY value {$sort} LIMIT 50", [$placeholder_name]);
```

Comment:
Will `${sort}` always be safe to insert into the query? Any chance it could be manipulated?

Reply:
Good point, added a check when we load placeholders from the DB to ensure it is always either ASC or DESC